<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (version 1.7.0_17) on Tue Feb 18 09:09:07 CET 2014 -->
<title>Uses of Package com.ohnosequences.bio4j.neo4j.model.relationships.features</title>
<meta name="date" content="2014-02-18">
<link rel="stylesheet" type="text/css" href="../../../../../../../stylesheet.css" title="Style">
</head>
<body>
<script type="text/javascript"><!--
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="Uses of Package com.ohnosequences.bio4j.neo4j.model.relationships.features";
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar_top">
<!--   -->
</a><a href="#skip-navbar_top" title="Skip navigation links"></a><a name="navbar_top_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li>Class</li>
<li class="navBarCell1Rev">Use</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev</li>
<li>Next</li>
</ul>
<ul class="navList">
<li><a href="../../../../../../../index.html?com/ohnosequences/bio4j/neo4j/model/relationships/features/package-use.html" target="_top">Frames</a></li>
<li><a href="package-use.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../../../../../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip-navbar_top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<div class="header">
<h1 title="Uses of Package com.ohnosequences.bio4j.neo4j.model.relationships.features" class="title">Uses of Package<br>com.ohnosequences.bio4j.neo4j.model.relationships.features</h1>
</div>
<div class="contentContainer">
<ul class="blockList">
<li class="blockList">
<table border="0" cellpadding="3" cellspacing="0" summary="Use table, listing packages, and an explanation">
<caption><span>Packages that use <a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/package-summary.html">com.ohnosequences.bio4j.neo4j.model.relationships.features</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Package</th>
<th class="colLast" scope="col">Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><a href="#com.ohnosequences.bio4j.neo4j.model.nodes">com.ohnosequences.bio4j.neo4j.model.nodes</a></td>
<td class="colLast">
<div class="block">General nodes</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#com.ohnosequences.bio4j.neo4j.model.relationships.features">com.ohnosequences.bio4j.neo4j.model.relationships.features</a></td>
<td class="colLast">
<div class="block">Protein features relationships</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#com.ohnosequences.bio4j.neo4j.programs">com.ohnosequences.bio4j.neo4j.programs</a></td>
<td class="colLast">
<div class="block">Bio4j importing process programs/modules</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="com.ohnosequences.bio4j.neo4j.model.nodes">
<!--   -->
</a>
<table border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/package-summary.html">com.ohnosequences.bio4j.neo4j.model.relationships.features</a> used by <a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/nodes/package-summary.html">com.ohnosequences.bio4j.neo4j.model.nodes</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/ActiveSiteFeatureRel.html#com.ohnosequences.bio4j.neo4j.model.nodes">ActiveSiteFeatureRel</a>
<div class="block">Amino acid(s) directly involved in the activity of an enzyme</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/SignalPeptideFeatureRel.html#com.ohnosequences.bio4j.neo4j.model.nodes">SignalPeptideFeatureRel</a>
<div class="block">Sequence targeting proteins to the secretory pathway or periplasmic space</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/SpliceVariantFeatureRel.html#com.ohnosequences.bio4j.neo4j.model.nodes">SpliceVariantFeatureRel</a>
<div class="block">Amino acid change(s) producing alternate protein isoforms</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/TransmembraneRegionFeatureRel.html#com.ohnosequences.bio4j.neo4j.model.nodes">TransmembraneRegionFeatureRel</a>
<div class="block">Extent of a membrane-spanning region</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="com.ohnosequences.bio4j.neo4j.model.relationships.features">
<!--   -->
</a>
<table border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/package-summary.html">com.ohnosequences.bio4j.neo4j.model.relationships.features</a> used by <a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/package-summary.html">com.ohnosequences.bio4j.neo4j.model.relationships.features</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/BasicFeatureRel.html#com.ohnosequences.bio4j.neo4j.model.relationships.features">BasicFeatureRel</a>
<div class="block">Basic feature relationship</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="com.ohnosequences.bio4j.neo4j.programs">
<!--   -->
</a>
<table border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/package-summary.html">com.ohnosequences.bio4j.neo4j.model.relationships.features</a> used by <a href="../../../../../../../com/ohnosequences/bio4j/neo4j/programs/package-summary.html">com.ohnosequences.bio4j.neo4j.programs</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/ActiveSiteFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">ActiveSiteFeatureRel</a>
<div class="block">Amino acid(s) directly involved in the activity of an enzyme</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/BindingSiteFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">BindingSiteFeatureRel</a>
<div class="block">Binding site for any chemical group (co-enzyme, prosthetic group, etc.)</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/CalciumBindingRegionFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">CalciumBindingRegionFeatureRel</a>
<div class="block">Position(s) of calcium binding region(s) within the protein</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/ChainFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">ChainFeatureRel</a>
<div class="block">Extent of a polypeptide chain in the mature protein</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/CoiledCoilRegionFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">CoiledCoilRegionFeatureRel</a>
<div class="block">Positions of regions of coiled coil within the protein</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/CompositionallyBiasedRegionFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">CompositionallyBiasedRegionFeatureRel</a>
<div class="block">Region of compositional bias in the protein</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/CrossLinkFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">CrossLinkFeatureRel</a>
<div class="block">Residues participating in covalent linkage(s) between proteins</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/DisulfideBondFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">DisulfideBondFeatureRel</a>
<div class="block">Cysteine residues participating in disulfide bonds</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/DnaBindingRegionFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">DnaBindingRegionFeatureRel</a>
<div class="block">Position and type of a DNA-binding domain</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/DomainFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">DomainFeatureRel</a>
<div class="block">Position and type of each modular protein domain</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/GlycosylationSiteFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">GlycosylationSiteFeatureRel</a>
<div class="block">Covalently attached glycan group(s)</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/HelixFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">HelixFeatureRel</a>
<div class="block">Helical regions within the experimentally determined protein structure</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/InitiatorMethionineFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">InitiatorMethionineFeatureRel</a>
<div class="block">Cleavage of the initiator methionine</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/IntramembraneRegionFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">IntramembraneRegionFeatureRel</a>
<div class="block">Extent of a region located in a membrane without crossing it</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/LipidMoietyBindingRegionFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">LipidMoietyBindingRegionFeatureRel</a>
<div class="block">Covalently attached lipid group(s)</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/MetalIonBindingSiteFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">MetalIonBindingSiteFeatureRel</a>
<div class="block">Binding site for a metal ion</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/ModifiedResidueFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">ModifiedResidueFeatureRel</a>
<div class="block">Modified residues excluding lipids, glycans and protein cross-links</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/MutagenesisSiteFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">MutagenesisSiteFeatureRel</a>
<div class="block">Site which has been experimentally altered by mutagenesis</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/NonConsecutiveResiduesFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">NonConsecutiveResiduesFeatureRel</a>
<div class="block">Indicates that two residues in a sequence are not consecutive</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/NonStandardAminoAcidFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">NonStandardAminoAcidFeatureRel</a>
<div class="block">Occurence of non-standard amino acids (selenocysteine and pyrrolysine) in the protein sequence</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/NonTerminalResidueFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">NonTerminalResidueFeatureRel</a>
<div class="block">The sequence is incomplete.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/NucleotidePhosphateBindingRegionFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">NucleotidePhosphateBindingRegionFeatureRel</a>
<div class="block">Nucleotide phosphate binding region</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/PeptideFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">PeptideFeatureRel</a>
<div class="block">Extent of an active peptide in the mature protein</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/PropeptideFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">PropeptideFeatureRel</a>
<div class="block">Part of a protein that is cleaved during maturation or activation</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/RegionOfInterestFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">RegionOfInterestFeatureRel</a>
<div class="block">Region of interest in the sequence</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/RepeatFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">RepeatFeatureRel</a>
<div class="block">Positions of repeated sequence motifs or repeated domains</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/SequenceConflictFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">SequenceConflictFeatureRel</a>
<div class="block">Description of sequence discrepancies of unknown origin</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/SequenceVariantFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">SequenceVariantFeatureRel</a>
<div class="block">Description of a natural variant of the protein</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/ShortSequenceMotifFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">ShortSequenceMotifFeatureRel</a>
<div class="block">Short (up to 20 amino acids) sequence motif of biological interest</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/SignalPeptideFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">SignalPeptideFeatureRel</a>
<div class="block">Sequence targeting proteins to the secretory pathway or periplasmic space</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/SiteFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">SiteFeatureRel</a>
<div class="block">Any interesting single amino acid site on the sequence</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/SpliceVariantFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">SpliceVariantFeatureRel</a>
<div class="block">Amino acid change(s) producing alternate protein isoforms</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/StrandFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">StrandFeatureRel</a>
<div class="block">Beta strand regions within the experimentally determined protein structure</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/TopologicalDomainFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">TopologicalDomainFeatureRel</a>
<div class="block">Location of non-membrane regions of membrane-spanning proteins</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/TransitPeptideFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">TransitPeptideFeatureRel</a>
<div class="block">Extent of a transit peptide for organelle targeting</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/TransmembraneRegionFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">TransmembraneRegionFeatureRel</a>
<div class="block">Extent of a membrane-spanning region</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/TurnFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">TurnFeatureRel</a>
<div class="block">Turns within the experimentally determined protein structure</div>
</td>
</tr>
<tr class="rowColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/UnsureResidueFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">UnsureResidueFeatureRel</a>
<div class="block">Regions of uncertainty in the sequence</div>
</td>
</tr>
<tr class="altColor">
<td class="colOne"><a href="../../../../../../../com/ohnosequences/bio4j/neo4j/model/relationships/features/class-use/ZincFingerRegionFeatureRel.html#com.ohnosequences.bio4j.neo4j.programs">ZincFingerRegionFeatureRel</a>
<div class="block">Position(s) and type(s) of zinc fingers within the protein</div>
</td>
</tr>
</tbody>
</table>
</li>
</ul>
</div>
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar_bottom">
<!--   -->
</a><a href="#skip-navbar_bottom" title="Skip navigation links"></a><a name="navbar_bottom_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../../../../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li>Class</li>
<li class="navBarCell1Rev">Use</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../../../../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../../../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../../../../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev</li>
<li>Next</li>
</ul>
<ul class="navList">
<li><a href="../../../../../../../index.html?com/ohnosequences/bio4j/neo4j/model/relationships/features/package-use.html" target="_top">Frames</a></li>
<li><a href="package-use.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../../../../../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip-navbar_bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
